internal/ociconv/ociconv.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+// Drops docker specific properties
+// See: https://github.com/opencontainers/image-spec/blob/main/config.md
+func toOCIV1Config(config v1.Config) v1.Config {
+	return v1.Config{
+		User:         config.User,
+		ExposedPorts: config.ExposedPorts,
+		Env:          config.Env,
+		Entrypoint:   config.Entrypoint,
+		Cmd:          config.Cmd,
+		Volumes:      config.Volumes,
+		WorkingDir:   config.WorkingDir,
+		Labels:       config.Labels,
+		StopSignal:   config.StopSignal,
+	}
+}
+
+func toOCIV1ConfigFile(cf *v1.ConfigFile) *v1.ConfigFile {
+	return &v1.ConfigFile{
+		Created:      cf.Created,
+		Author:       cf.Author,
+		Architecture: cf.Architecture,
+		OS:           cf.OS,
+		OSVersion:    cf.OSVersion,
+		History:      cf.History,
+		RootFS:       cf.RootFS,
+		Config:       toOCIV1Config(cf.Config),
+	}
+}
+
+// OCIImage mutates the provided v1.Image to be OCI compliant v1.Image
+// Check image-spec to see which properties are ported and which are dropped.
+// https://github.com/opencontainers/image-spec/blob/main/config.md
+func OCIImage(base v1.Image) (v1.Image, error) {
+	// Get original manifest
+	m, err := base.Manifest()
+	if err != nil {
+		return nil, err
+	}
+	// Convert config
+	cfg, err := base.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	cfg = toOCIV1ConfigFile(cfg)
+
+	layers, err := base.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	newLayers := []v1.Layer{}
+
+	for _, layer := range layers {
+		mediaType, err := layer.MediaType()
+		if err != nil {
+			return nil, err
+		}
+		switch mediaType {
+		case types.DockerLayer:
+			reader, err := layer.Compressed()
+			if err != nil {
+				return nil, fmt.Errorf("getting layer: %w", err)
+			}
+			layer, err = tarball.LayerFromReader(reader, tarball.WithMediaType(types.OCILayer))
+			if err != nil {
+				return nil, fmt.Errorf("building layer: %w", err)
+			}
+		case types.DockerUncompressedLayer:
+			reader, err := layer.Uncompressed()
+			if err != nil {
+				return nil, fmt.Errorf("getting layer: %w", err)
+			}
+			layer, err = tarball.LayerFromReader(reader, tarball.WithMediaType(types.OCIUncompressedLayer))
+			if err != nil {
+				return nil, fmt.Errorf("building layer: %w", err)
+			}
+		}
+		newLayers = append(newLayers, layer)
+	}
+
+	base, err = mutate.AppendLayers(empty.Image, newLayers...)
+	if err != nil {
+		return nil, err
+	}
+
+	base = mutate.MediaType(base, types.OCIManifestSchema1)
+	base = mutate.ConfigMediaType(base, types.OCIConfigJSON)
+	base = mutate.Annotations(base, m.Annotations).(v1.Image)
+	base, err = mutate.ConfigFile(base, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return base, nil
+}
+
+// OCIImageIndex mutates the provided v1.ImageIndex to be OCI compliant v1.ImageIndex
+func OCIImageIndex(base v1.ImageIndex) (v1.ImageIndex, error) {
```

Comment:
Does the image have to be OCI-ified as part of this? If not, I think we should punt on OCI-fication methods until a future change.

If it is required, I might still want to add it in a later PR so we can focus on getting it right. I've tried and failed to do this before, and the way this manifests is weird registry-specific failures that aren't easy to test here in isolation.

Reply:
I think this is a sort of relict from my previous tests. I got the source code for the OCI conversion stuff from this PR https://github.com/google/go-containerregistry/pull/1293. I added the index conversion as well, because I was not sure how a DockerManifestV2 Index List handles OCI Images, but it seems like to work.

I'll remove the unnecessary functions.